File: packages/coding-agent/src/tui/visual-truncate.ts
import { Text } from "@kennyfrc/mu-tui";

export interface VisualTruncateResult {
	/** Visual lines to display (already wrapped/padded to width) */
	visualLines: string[];
	/** Count of visual lines skipped (hidden) */
	skippedCount: number;
}

/**
 * Truncate text to a maximum number of *visual* lines (accounting for wrapping).
 *
 * This is especially important for tool output where a single logical line
 * (no newlines) might wrap into hundreds of terminal lines.
 *
 * Behavior: returns the last N visual lines (most recent output).
 */
export function truncateToVisualLines(
	text: string,
	maxVisualLines: number,
	width: number,
	paddingX = 0,
): VisualTruncateResult {
	if (!text) {
		return { visualLines: [], skippedCount: 0 };
	}

	const tempText = new Text(text, paddingX, 0);
	const allVisualLines = tempText.render(width);

	if (allVisualLines.length <= maxVisualLines) {
		return { visualLines: allVisualLines, skippedCount: 0 };
	}

	return {
		visualLines: allVisualLines.slice(-maxVisualLines),
		skippedCount: allVisualLines.length - maxVisualLines,
	};
}

File: python/src/pi_mono/coding_agent/modes/interactive/components/visual_truncate.py
"""Truncate text to a maximum number of visual lines."""

from __future__ import annotations

from dataclasses import dataclass

from pi_mono.tui.components.text import Text


@dataclass
class VisualTruncateResult:
    visual_lines: list[str]
    skipped_count: int


def truncate_to_visual_lines(
    text: str,
    max_visual_lines: int,
    width: int,
    padding_x: int = 0,
) -> VisualTruncateResult:
    if not text:
        return VisualTruncateResult(visual_lines=[], skipped_count=0)
    temp_text = Text(text, padding_x, 0)
    all_visual_lines = temp_text.render(width)
    if len(all_visual_lines) <= max_visual_lines:
        return VisualTruncateResult(visual_lines=all_visual_lines, skipped_count=0)
    truncated_lines = all_visual_lines[-max_visual_lines:]
    return VisualTruncateResult(
        visual_lines=truncated_lines,
        skipped_count=len(all_visual_lines) - max_visual_lines,
    )
